Completing the CFF Table

camertron opened the pull request:
_**This pull request is part of a larger effort to bring OTF support to TTFunk. See https://github.com/prawnpdf/ttfunk/issues/53 for details.**_

It's finally here! Apologies for the delay, and thank you for your patience, @pointlessone. This one's a beast.

Things to note in this PR:
1. All remaining CFF functionality has been finished, including support for charsets, encoding, charstrings, subroutines, and the top index.
2. It is still not possible to encode/subset a loaded font in opentype format. That's what the next PR is all about (it introduces the `OTFEncoder` class). Don't be looking for that functionality just yet.
3. There are a few `@TODO` annotations in the comments that indicate opportunities to improve on the existing functionality, but are not very important or are optimizations. For example, we currently don't detect unused subroutines for the given subset.
4. The `Index` data structure has been updated to use `EncodedString`s again instead of `.map { ... }.join` because it can be composed of other data structures (usually dicts) that contain offsets to other tables and therefore need to take advantage of placeholders.
5. I had a really hard time finding examples of fonts that demonstrate certain CFF features. In such cases I have attempted to test those features in creative ways as well as leave comments behind.
6. I have included the binary versions of several freely available fonts, and, in one case, the corresponding license file. New fonts include ComicJens-Regular, Exo-Regular, and AlbertTextBold.
7. Charstrings (the CFF version of glyphs) can be rendered in SVG, which I think is pretty cool  😎

pointlessone commented:
@camertron Sorry for the delay. I have started reviewing it a while back. It's a rather big PR and I'm trying to validate it against the spec as well. Please bare with me.

camertron commented:
No problem @pointlessone, thanks for your diligence!

pointlessone commented:
@camertron I emailed you a while back. I wonder if you got the email.

camertron commented:
@pointlessone Yes, apologies for not responding in a timely manner. I just sent you a response :)

camertron commented on lib/ttfunk/bin_utils.rb:
```diff
@@ -31,6 +31,22 @@ def twos_comp_to_int(num, bit_width:)
+        end
+
+        ranges << [start, values.last - start]
+      end
```
Ok, I can change this. It looks like `slice_when` was introduced in ruby 2.2, is it ok to not support 2.1? The travis.yml file would suggest it is, but I wanted to make sure.

pointlessone replied:
Our official stance it to support supported Rubies. That means that we should drop 2.2 here as we did in the main Prawn repo.

camertron replied:
👍

camertron commented on lib/ttfunk/table/cff.rb:
```diff
@@ -1,27 +1,63 @@
+            top_index.encode { |top_dict| top_dict.encode(mapping) },
+            string_index.encode,
+            global_subr_index.encode
+          ]
```
Right, those are all encoded (if they exist) with the top dict (i.e. when `top_dict.encode(mapping)` is invoked).

pointlessone replied:
Is there any specific reason we can't have them here? In the spec they seem to be on about the same level in hierarchy.

camertron replied:
Yeah, the top dict holds relative offsets to all of these additional data structures, meaning 1) the structures cannot come before the top dict (since offsets are unsigned), and 2) the structures are logically part of the top dict, which in my opinion means the top dict class should have control over where they are written. While it's not an absolute requirement the structures be encoded by the top dict itself, I think it would be awkward to ask the top dict for them and encode them after the top dict. The `TopDict` class would need to create several placeholders within the encoded string which would then be resolved by the `CFF` class. In my mind that violates SRP.

camertron commented on lib/ttfunk/table/cff.rb:
```diff
@@ -1,27 +1,63 @@
+      autoload :PrivateDict,      'ttfunk/table/cff/private_dict'
+      autoload :SubrIndex,        'ttfunk/table/cff/subr_index'
+      autoload :TopDict,          'ttfunk/table/cff/top_dict'
+      autoload :TopIndex,         'ttfunk/table/cff/top_index'
```
This is a very common misconception. `autoload` has been thread-safe since ruby 1.9.3. It is certainly thread-safe in ruby 2.0 and later. See: https://bugs.ruby-lang.org/issues/921

pointlessone replied:
Good news! I'm glad this issue got some attention. However it seem the feature itself is still problematic:
* Here's a relatively recent issue [#14742](https://bugs.ruby-lang.org/issues/14742). It's resolved but illustrates an old maxim of "threads are hard".
* Here's an older issue [#12688](https://bugs.ruby-lang.org/issues/12688) that haven't been resolved yet.

What I'm trying to say is that gain is marginal but potential for issues is considerable. I'd still avoid autoload if possible.

camertron replied:
Hmm that's definitely food for thought. However, I would say the issues you linked to are fairly edge case-y. TTFunk generally doesn't define multiple classes per file (we should fix places where it does), so the first issue where multiple constants are autoloaded from the same file shouldn't be a problem. The second issue where a class is re-opened and asked to re-autoload the same constant over and over again I would think is also a pretty rare problem.

I've been using autoload in my own libraries for years and never had problems with it. Moreover, autoload is used in bundler, rack, and sinatra, some of the most popular and used Ruby libraries.

In my opinion, the main benefit of autoload is to avoid having to require your entire library when your application (or whatever) starts up. To me it's very convenient to be able to say `require 'ttfunk'` and then be able to reference any constant within the `TTFunk` namespace. As it is, I have to specifically require `ttfunk/subset` in order to access the `TTFunk::Subset` class, which I find sort of annoying.

camertron commented on lib/ttfunk/table/cff/charset.rb:
```diff
@@ -0,0 +1,230 @@
+            @entries = []
+            @length = 0
+
+            until @count >= top_dict.charstrings_index.count - 1
```
Hmm so what should we do? Since the number of ranges isn't specified, how will we know if we read past the end of the range list?

pointlessone replied:
Canonically, we'd at least had `eof?`. We can check that and maybe bounds of the table we're parsing.

camertron replied:
Ok that makes sense 👍

pointlessone replied:
Can we do anything about this? Or do you think it's out of scope?

camertron replied:
Ah looks like I forgot about this one too. I just pushed a potential solution, let me know what you think :)

camertron commented on lib/ttfunk/table/cff/charset.rb:
```diff
@@ -0,0 +1,230 @@
+            @entries = []
+            @length = 0
+
+            until @count >= top_dict.charstrings_index.count - 1
```
I have long wondered why this isn't a pattern already present in TTFunk. Instead all we really have is the `Reader` module which exposes `read` and friends. It would be really nice to be able to define methods like `read_ushort` and give the object some start position and max length. I would think that's something we could file an issue for and implement as a separate chunk of work since it would require a fairly large effort to put into practice across the whole library.

camertron commented on lib/ttfunk/table/cff/charset.rb:
```diff
@@ -0,0 +1,230 @@
+
+        def find_string(sid)
+          if offset
+            return self.class.standard_strings[sid - 1] if sid <= 390
```
Hmm that might be a good idea, hopefully I can avoid additional off-by-one issues when adding it to the various encodings 😅

camertron replied:
This one was pretty difficult to address because the fix isn't as simple as adding `.notdef` to each predefined charset. As it happens, the strings index also suffers from off by one issues because it too excludes GID 0. Since the strings index isn't an array though, prepending `.notdef` isn't easy to do like it is for the predefined charsets. The documentation assumes you understand the implicitness of GID 0 and would I think be more difficult to square up with what's in the code if we made allowances by inserting `.notdef` ourselves. That said, I do take your point that the many small adjustments are more prone to error, so I created a couple of classes to help abstract them away. I realized the majority of the discrepancies come from the fact that the predefined charsets and string index are really just 1-based arrays. The new class `OneBasedArray` and `OneBasedIndex` wrap `Array` and `CFF::Index` respectively and allow us to access elements directly by GID. Let me know what you think :)

camertron commented on lib/ttfunk/table/cff/charstrings_index.rb:
```diff
@@ -0,0 +1,37 @@
+        end
+
+        def [](index)
+          entry_cache[index] ||= TTFunk::Table::Cff::Charstring.new(
```
That would be great, except for the unfortunate fact that we need to know the dimensions of the glyphs in order to calculate entries in the hmtx and maxp tables. Without calculating the starting/ending coordinates of all the contours, I'm not sure that information is derivable in any other way.

pointlessone replied:
Are you saying that there are no glyph metrics anywhere? Do I read it correctly? How do you specify width of a space, for instance? It usually doesn't have any contours. I imagine, glyphs that are used for compositing (e.g. accents, graves, etc.) are supposed to have some space beyond their contours.

We do this with TTF: parse just enough to get font metrics and perform subsetting but not all the way to the contours. I was under impression we can do that with OTF as well.

camertron replied:
Well, the hmtx table includes glyph metrics for both font flavors, but it only includes a couple of the ones we need (advance width and left side bearing). The `glyf` table in PostScript-flavored fonts (i.e. TrueType fonts) includes the font's bounding box in the header entry for each glyph, which can be parsed out separately from all the other glyph information. CFF-flavored fonts however don't have a `glyf` table and don't explicitly define bounding boxes. I believe the width of a space is defined by the font's default glyph width, which is stored in the private dict.

I could be totally wrong about this, but I haven't read anything to the contrary in the spec or during the course of my research into CFF fonts.

camertron replied:
Were you able to find any additional information about CFF-flavored font metrics?

pointlessone replied:
I took a cursory look and it appears parsing charstrings is the only way to get proper glyph metrics. We have no choice here.

camertron commented on lib/ttfunk/table/cff/charstring.rb:
```diff
@@ -0,0 +1,489 @@
+          @font_dict = font_dict
+          @raw = raw
+
+          @default_width_x = (@font_dict || @top_dict)
```
Ok, so the `Charstring` class commits a fair number of heinous ruby crimes in order to maximize performance. I was able to dramatically decrease the time it takes to subset the Noto Sans font (tens of thousands of glyphs) by using ivars instead of `attr_readers`. The function call overhead really makes a difference.

pointlessone replied:
In this particular instance I meant it can be a local variable since it doesn't seem to be used outside of this method.

camertron replied:
Oh hehe ok 👍

pointlessone replied:
Could you please address this one?

camertron replied:
Oops sorry. It's done now.

pointlessone replied:
Can't see it. Have you forgot to push the change?

camertron replied:
Lol yes I did. It's pushed now.

pointlessone commented on lib/ttfunk/table/cff/charset.rb:
```diff
@@ -0,0 +1,230 @@
+          def strings_for_charset_id(charset_id)
+            string_cache[charset_id] ||= YAML.load_file(
+              ::File.expand_path(
+                ::File.join('.', 'charsets', CHARSET_FILES.fetch(charset_id)),
```
Let's go the Ruby route.

camertron replied:
👍

camertron commented on lib/ttfunk/table/cff/charset.rb:
```diff
@@ -0,0 +1,230 @@
+          case format_sym
+          when :array_format
+            # zero is always .notdef, so adjust with - 1
+            @entries[glyph_id - 1]
```
Several points to bring up here:

1. I tried adding `attr_writer`s but they felt awkward to use in assignments because they look like local variables, which strikes me as the same sort of error-proneness as ivar typos. I think I would prefer to always use ivars when assigning and `attr_reader`s when accessing, and keep assignments localized to the constructor (and in this case the `parse` method, which can be thought of as a constructor of sorts for these subtables).
2. Even if we decided to avoid ivar usage completely during assignment, one of the ivars I'm using here is `format` which is also a method on `Kernel`. Sure, we could shadow it with an `attr_reader`/`attr_writer` but that feels... wrong. Maybe I should just rename it?

pointlessone replied:
One way you can make assignments more explicit is to use `self`:
```ruby
self.entries = []
```

While this doesn't fundamentally fix the issue it can help readers distinguish local vars from attributes.

I'll leave it up to you.

camertron commented:
Ok @pointlessone I think I've addressed all your comments so far. Let me know how you'd like to proceed :)

pointlessone commented on lib/ttfunk/table/cff/charsets/expert.rb:
```diff
@@ -0,0 +1,387 @@
+      module Charsets
+        EXPERT = OneBasedArray.new([
+          'space',
+          nil,
```
You can save a couple of lines by using the following trick:
```
[
  * [nil] * 11
]
```

It's a combination of two separate things.
1. Array multiplication: `[nil] * 3` gives you `[nil, nil, nil]`
2. Splat operator. It's often used in method calls but also works in array literals. It basically unwrapps an array into a series of values: `[* [1, 2, 3]]` is the same as `[1, 2, 3]`.

I don't insist as these files are not going to be edited often. Mention because it's a neat trick and might be useful in cases like this.

camertron replied:
Very cool, thank you!

pointlessone commented:
@camertron I've found a few minutes to resolve addressed discussions. Could you please update me on the remaining ones?

camertron commented:
@pointlessone I think the only remaining issue is what to do about the off-by-one problems. I like the simplicity of `OneBasedArray` but I also understand your concerns about inheriting from `Array`. I'll work on an update to wrap `Array`instead as you suggested.

Oh also, in not sure we ever reached consensus about what to do about `autoload`. Other than that, I think everything's been addressed 😊

pointlessone commented:
We already have autoload in the lib. Let’s keep it. I’ll research it a bit more and remove it if I think it’s still problematic later. Outside of the OTF work.

camertron commented:
@pointlessone ok, I fixed the `Array` inheritance issue and replaced runs of `nil`s and `0`s with the splat syntax you suggested. I think that's everything :)

camertron commented:
FYI @pointlessone you'll probably want to merge [this guy](https://github.com/prawnpdf/ttfunk/pull/65) into `otf` before merging `cff_rest`.

pointlessone commented on spec/ttfunk/table/cff/charset_spec.rb:
```diff
@@ -14,7 +14,7 @@
       it 'includes all the entries from the expert charset' do
         expect(charset.to_a).to eq(
-          ['.notdef'] + described_class.strings_for_charset_id(charset_id)
+          ['.notdef'] + described_class.strings_for_charset_id(charset_id).to_a
```
You can spare explicit `to_a` if you implement `to_ary` method on `OneBasedArray`.

Example:
```ruby
class A
  def to_ary
    [:A]
  end
end

[1] + A.new # => [1, :A]
```

camertron replied:
Wow I didn't know about that trick, thanks!

pointlessone commented:
Aw… CI fails. Maybe we'll have to rebase sooner.

camertron commented:
Yeah that's originally why I submitted the PR to update the `otf` branch 😅

pointlessone commented:
The PR looks fine. And since it's not going straight into master we can deal with CI after it's merged.

camertron commented:
@pointlessone how would you feel about merging this guy?

pointlessone commented:
I will rebase `otf` branch over the weekend.

Thank you for this massive work!

pointlessone commented:
@camertron `otf` brunch's updated. It was rebased on master so you may want to pull a fresh copy from the repo for your next PRs.
